### PR TITLE
Remove NU5105, do not warn when a package with a semver2 version is created

### LINKED ISSUE #5201: nuget.exe pack should no longer warn about packing a SemVer 2.0.0 package
Consider this nuspec:

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
    <metadata>
        <id>jver.bugbash.dependency</id>
        <version>1.0.0-b1</version>
        <authors>jver</authors>
        <requireLicenseAcceptance>false</requireLicenseAcceptance>
        <description>My package description.</description>
        <dependencies>
            <dependency id="Newtonsoft.Json" version="1.0.0-alpha.1" />
        </dependencies>
    </metadata>
</package>
```

Command: `nuget.exe pack jver.bugbash.dependency.nuspec`
```
Attempting to build package from 'jver.bugbash.dependency.nuspec'.
Successfully created package 'C:\Users\jver\Desktop\packages\jver.bugbash.dependency.1.0.0\jver.bugbash.dependency.1.0.0-b1.nupkg'.
```

This is a SemVer 2.0.0 package because the dependency version range has a dot in it, i.e. `1.0.0-alpha.1`.

However `nuget.exe pack` does not warn about this SemVer 2.0.0-ness like it does when the package version itself has a dot.

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
    <metadata>
        <id>jver.bugbash.dependency</id>
        <version>1.0.0-b.1</version>
        <authors>jver</authors>
        <requireLicenseAcceptance>false</requireLicenseAcceptance>
        <description>My package description.</description>
    </metadata>
</package>
```

Command: `nuget.exe pack jver.bugbash.dependency.nuspec`
```
Attempting to build package from 'jver.bugbash.dependency.nuspec'.
Successfully created package 'C:\Users\jver\Desktop\packages\jver.bugbash.dependency.1.0.0\jver.bugbash.dependency.1.0.0-b.1.nupkg'.
WARNING: Issue found with package 'jver.bugbash.dependency'.

WARNING: Issue: Package version not supported on legacy clients.
WARNING: Description: The package version '1.0.0-b.1' uses SemVer 2.0.0 or components of SemVer 1.0.0 that are not supported on legacy clients. This message can be ignored if the package is not intended for older clients.
WARNING: Solution: Change the package version to a SemVer 1.0.0 string. If the version contains a release label it must start with a letter.
```

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/5201
Regression: Yes/No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
Related to https://github.com/NuGet/NuGetGallery/issues/7836
@joelverhagen perfectly summarized it there. Rephrasing (copying) most of his description.

**Note**: The nuget.org equivalent work is already being done in https://github.com/NuGet/NuGetGallery/pull/7837, so it'd be nice for both the client and the server make the change around the same time. 

When the SemVer 2.0.0 changes were introduced to the V2 and V3 protocol, they are intended to mitigate the impact of SemVer 2.0.0 packages breaking old clients, e.g. suppose a very popular package released a SemVer 2.0.0 version. In such a case a 2.x customer would fail to restore a package because the version parsing would explode. So we make SemVer 2.0.0 packages "opt-in". 

This protocol change meant that SemVer 2.0.0 packages were hidden in most of our APIs for clients that predate NuGet version 4.3. 4.3 is when the SemVer 2.0.0 protocol was implemented. 
At that time a pack warning was added.

Since 4.3 was shipped so long ago, most clients now support the SemVer 2.0.0 protocol and this warning is more noisy than helpful.

Range | Percent | Note
-- | -- | --
≥ 4.3 | ~96.4% | Supports SemVer 2.0.0 protocol
≥ 3.0 && < 4.3 | ~3.6% | Partial SemVer 2.0.0 but not SemVer 2.0.0 protocol `*`
< 3.0 | ~0.0% | Does not support SemVer 2.0.0

`*` =
- Anything before 4.3 will not see SemVer 2.0.0 in search or in the PM UI.
- 3.5.0 and on supported SemVer 2.0.0 restore in packages.config so people can hand edit to work-around.
- PackageReference always supported SemVer 2.0.0 from what I can tell so people could hard-code a SemVer 2.0.0 package in their .csproj and still restore.
- V3 feeds always return SemVer 2.0.0 packages from flat container so packages can be found prior to SemVer 2.0.0. protocol support

The biggest motivation here is that needless warnings often force customers to no warn or in extreme scenario run NoPackageAnalysis=true which renders all of our current/future package quality warning useless. 

In this PR, I am removing any reference to the warning itself. I don't see a reason to keep the code around. 

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/Rules/ContentFolderInPackageReferenceProjectRule.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 namespace NuGet.Packaging.Rules
 {
-    class ContentFolderInPackageReferenceProjectRule : IPackageRule
+    internal class ContentFolderInPackageReferenceProjectRule : IPackageRule
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/ContentFolderInPackageReferenceProjectRule.cs:9]: As per [doc](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers#class-and-struct-accessibility), `Internal` is default access modifier for Classes and structs that are declared directly within a namespace.

File: src/NuGet.Core/NuGet.Packaging/Rules/InstallScriptInPackageReferenceProjectRule.cs [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 namespace NuGet.Packaging.Rules
 {
-    class InstallScriptInPackageReferenceProjectRule : IPackageRule
+    internal class InstallScriptInPackageReferenceProjectRule : IPackageRule
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/InstallScriptInPackageReferenceProjectRule.cs:9]: see this [comment](https://github.com/NuGet/NuGet.Client/pull/3226#discussion_r377183570)

File: src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs [hunk 3]
```diff
@@ -2,24 +2,17 @@
...
 using NuGet.Common;
-using NuGet.ContentModel;
 using NuGet.Frameworks;
-using NuGet.Packaging.Core;
-using NuGet.RuntimeModel;
 
 namespace NuGet.Packaging.Rules
 {
-    class ReferencesInNuspecMatchRefAssetsRule : IPackageRule
+    internal class ReferencesInNuspecMatchRefAssetsRule : IPackageRule
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs:15]: see this [comment](https://github.com/NuGet/NuGet.Client/pull/3226#discussion_r377183570)

File: src/NuGet.Core/NuGet.Packaging/Rules/UnrecognizedLicenseIdentifierRule.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
 namespace NuGet.Packaging.Rules
 {
-    class UnrecognizedLicenseIdentifierRule : IPackageRule
+    internal class UnrecognizedLicenseIdentifierRule : IPackageRule
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Packaging/Rules/UnrecognizedLicenseIdentifierRule.cs:13]: see this [comment](https://github.com/NuGet/NuGet.Client/pull/3226#discussion_r377183570)

File: src/NuGet.Core/NuGet.Packaging/Rules/ContentFolderInPackageReferenceProjectRule.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Rules/ContentFolderInPackageReferenceProjectRule.cs:9]: [Rule #5](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md)

File: src/NuGet.Core/NuGet.Packaging/Rules/InstallScriptInPackageReferenceProjectRule.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Rules/InstallScriptInPackageReferenceProjectRule.cs:9]: [Rule #5](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md)

File: src/NuGet.Core/NuGet.Packaging/Rules/UnrecognizedLicenseIdentifierRule.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Rules/UnrecognizedLicenseIdentifierRule.cs:13]: [Rule #5](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md)

File: src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs:15]: [Rule #5](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md)

**Comment by nkolev92** [comment]
🔔

File: src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj [hunk 5]
```diff
@@ -121,16 +121,28 @@
     </EmbeddedResource>
   </ItemGroup>
 
-  <Target Name="ILMergeNuGetExe" AfterTargets="Build" Condition="'$(BuildingInsideVisualStudio)' != 'true'">
+  <!-- 	Since we are moving some code and strings from NuGet.CommandLine to NuGet.Commands, we opted to go through normal localization process
+        (build .resources.dll) and then add them to the ILMerged nuget.exe -->
+  <!-- This will also be called from CI build, after assemblies are localized, since our test infra takes nuget.exe before Localization  -->
+  <Target Name="ILMergeNuGetExe" AfterTargets="Build" Condition="'$(BuildingInsideVisualStudio)' != 'true' and '$(SkipILMergeOfNuGetExe)' != 'true'" >
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj]: Why is GitHub showing these changes? It looks like from Rob's `dotnet nuget ... sources` PR, and completely unrelated to NU5105.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs [hunk 6]
```diff
@@ -2715,26 +2713,29 @@ public void PackCommand_NoWarn_SuppressesWarnings(string property, string value,
...
                 Assert.True(File.Exists(nupkgPath), "The output .nupkg is not in the expected place");
                 Assert.True(File.Exists(nuspecPath), "The intermediate nuspec file is not in the expected place");
-                var expectedWarning = string.Format("warning " + NuGetLogCode.NU5105 + ": " + NuGet.Packaging.Rules.AnalysisResources.LegacyVersionWarning, semver2Version);
-                Assert.Equal(result.AllOutput.Contains(expectedWarning), expectToWarn);
-
+                var expectedWarning = string.Format("warning " + NuGetLogCode.NU5125 + ": " + NuGet.Packaging.Rules.AnalysisResources.LicenseUrlDeprecationWarning);
+                if (expectToWarn)
+                {
+                    result.AllOutput.Should().Contain(expectedWarning);
+                }
```
Comment (zivkan) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:2737]: This has a different semantic meaning. Previously, when the message was not expected, it asserted that the message was not presented. Now, it only checks that the message is present when it's expected, but does not validate when the message is not expected.

File: src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj]: Fixed.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs (hunk 6 above)
Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:2737]: yup, nice catch :)

Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:2737]: fixed.

**Review by zivkan — APPROVED** [review]

**Comment by nkolev92** [comment]
Had some test failures that should be fixed. 
Let me know if anyone wants to review this beside Andy, otherwise I'll merge Tuesday.

**Pull request merged**